blockstack/stacks-blockchain
# Feature/net p2p pingback

This PR adds two new features to the peer-to-peer networking code: **pingbacks** and **inbound neighbor walks**.

Pingbacks are when a node tries to handshake with another node that has recently handshaked with it.  This causes each publicly-routable node to add a directed edge to all other publicly-routable peers that talk to it (i.e. who are the origins of the node's inbound edges), thereby ensuring that there exists a bi-directional edge in the peer graph between any pair (A, B) of publicly routable nodes such that A connects to B _or_ B connects to A.

Inbound neighbor walks are when a node asks a peer with an inbound edge for its neighbors, and then proceeds to try to handshake with them.  This causes nodes that are publicly routable to ask nodes that may _not_ be publicly routable for their sets of publicly-routable nodes, thereby ensuring that there will exist a bi-directional edge in the peer graph between any pair (A, B) of publicly routable nodes if there exists distinct node C such that edges `C --> A` and `C --> B` exist (the existence of these edges will cause `A --> B` and `B --> A` to come into existence once A or B does an inbound neighbor walk to C).

If a node's maximum number of inbound and outbound edges are infinite, then these changes ensure that if the initial peer graph is connected, then the peer graph will evolve to the point where there exists a bidirectional edge between every publicly-routable node, and a directional edge from each non-publicly-routable node (i.e. those behind NATs) to every publicly routable node.  The graph would look like two concentric rings -- a "core" ring of publicly-routable nodes that are all-to-all connected, and a "peripheral" ring of non-publicly-routable nodes that are connected to all publicly-routable nodes.  This is the ideal network topology we want.

In practice, the number of inbound and outbound edges are capped, so the resulting peer graph will consist of the same vertices (nodes), but with a random subset of the edges.

EDIT: this PR updates `GET /v2/neighbors` to return not only a random frontier sample, but also a list of inbound and outbound peers.  This should make it easier external tooling to monitor the status of the peer network as we're rolling it out.

EDIT: this PR also fixes #1581 by adding a `HttpRequestType::Unmatched()` variant to capture unmatched paths (which the RPC server handles with a `HttpResponsType::NotFound()`).

## How to Test

To test the inbound neighbor walk feature, you'd spin up two nodes: a "remote" node that your laptop can contact (doesn't have to be publicly-routable), and a "local" node that has both this "remote" node and the Neon master node as its initial peers.  You should see the "remote" node ask the "local" node for neighbors, and in so doing, discover the Neon master node.  You may or may not see a pingback occur if you look at the logs.
